### ISSUE TITLE
`Heap::dealloc` is unsound

### ISSUE THREAD

**Issue by nikitax44** [rcore-os/buddy_system_allocator#37]
it is possible to cause segfault in entirely safe code:
```rs
fn main() {
    let mut heap = buddy_system_allocator::Heap<10>::new();
    heap.dealloc(
        std::ptr::NonNull::dangling(),
        std::alloc::Layout::from_size_align(1, 1).unwrap(),
    );
}
```

**Comment by jiegec** [comment]
Yes, but as an allocator, we don't want to support such case, since the expected usage is via global_allocator.

**Comment by nikitax44** [comment]
this function should be `unsafe`. you shouldn't expose safe functions that may cause UB. the `unsafe` attribute does not conflict with global allocator use case as `GlobalAlloc::dealloc` is declared to be `unsafe`.

**Comment by jiegec** [comment]
> this function should be `unsafe`. you shouldn't expose safe functions that may cause UB. the `unsafe` attribute does not conflict with global allocator use case as `GlobalAlloc::dealloc` is declared to be `unsafe`.

Okay, please add the unsafe attribute in a pull request.
